File: packages/swift/Sources/B1CodesLoaders/Loaders/B1CodesLoaderController.swift
import Foundation
import Combine

/// Imperative controller for B1CodesLoader SwiftUI views.
///
/// Allows consuming code to trigger `resolve()` exit discharge sequences,
/// or `pause()` / `resume()` particle simulations.
public class B1CodesLoaderController: ObservableObject {
    @Published public var isPaused: Bool = false
    @Published public var isResolving: Bool = false

    public init() {}

    /// Triggers the Thermal Glow exit discharge sequence (~350ms),
    /// accelerating particles outward before dissolving and firing `onDismiss`.
    public func resolve() {
        guard !isResolving else { return }
        isResolving = true
    }

    /// Pauses the particle simulation animation loop.
    public func pause() {
        guard !isPaused else { return }
        isPaused = true
    }

    /// Resumes the particle simulation animation loop.
    public func resume() {
        guard isPaused else { return }
        isPaused = false
    }

    /// Resets controller state to initial.
    public func reset() {
        isPaused = false
        isResolving = false
    }
}

File: packages/flutter/lib/src/loaders/b1codes_loader_controller.dart
import 'package:flutter/foundation.dart';

/// Imperative controller for b1codes particle loaders.
///
/// Allows consuming code to trigger [resolve] exit discharge sequences,
/// or [pause] / [resume] particle simulations.
class B1CodesLoaderController extends ChangeNotifier {
  bool _isPaused = false;
  bool _isResolving = false;

  bool get isPaused => _isPaused;
  bool get isResolving => _isResolving;

  /// Triggers the Thermal Glow exit discharge sequence (~350ms),
  /// accelerating particles outward before dissolving and firing `onDismiss`.
  void resolve() {
    if (_isResolving) return;
    _isResolving = true;
    notifyListeners();
  }

  /// Pauses the particle simulation animation loop.
  void pause() {
    if (_isPaused) return;
    _isPaused = true;
    notifyListeners();
  }

  /// Resumes the particle simulation animation loop.
  void resume() {
    if (!_isPaused) return;
    _isPaused = false;
    notifyListeners();
  }

  /// Reset controller state.
  void reset() {
    _isPaused = false;
    _isResolving = false;
    notifyListeners();
  }
}
